stan/math/opencl/matrix_cl.hpp

```diff
@@ -414,63 +412,43 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
   /**
    * Assign a \c matrix_cl to another
    */
-  matrix_cl<T>& operator=(const matrix_cl<T>& a) {
+  matrix_cl<T>& operator=(matrix_cl<T>&& a) {
     if (a.size() == 0) {
       return *this;
     }
     view_ = a.view();
     rows_ = a.rows();
     cols_ = a.cols();
     this->wait_for_read_write_events();
-    cl::CommandQueue queue = opencl_context.queue();
-    try {
-      cl::Event copy_event;
-      queue.enqueueCopyBuffer(a.buffer(), this->buffer(), 0, 0,
-                              a.size() * sizeof(T), &a.write_events(),
-                              &copy_event);
-      this->add_write_event(copy_event);
-      a.add_read_event(copy_event);
-    } catch (const cl::Error& e) {
-      check_opencl_error("copy (OpenCL)->(OpenCL)", e);
-    }
-    return *this;
-  }
-
-  /**
-   * Move a \c matrix_cl to another
-   */
-  matrix_cl<T>& operator=(matrix_cl<T>&& a) {
-    // Need to wait for all of matrices events before destroying old buffer
-    this->wait_for_read_write_events();
-    buffer_cl_ = a.buffer();
-    view_ = a.view();
+    buffer_cl_ = std::move(a.buffer_cl_);
     write_events_ = std::move(a.write_events_);
     read_events_ = std::move(a.read_events_);
-    rows_ = a.rows();
-    cols_ = a.cols();
     return *this;
   }
 
   /**
-   * Assign a \c matrix_cl of one arithmetic type to another
+   * Assign a \c matrix_cl to another
    */
-  template <typename U, typename = enable_if_arithmetic<U>>
-  matrix_cl<T>& operator=(const matrix_cl<U>& a) {
-    // Need to wait for all of matrices events before destroying old buffer
-    this->wait_for_read_write_events();
-    buffer_cl_ = a.buffer();
+  matrix_cl<T>& operator=(const matrix_cl<T>& a) {
+    if (a.size() == 0) {
+      return *this;
+    }
     view_ = a.view();
     rows_ = a.rows();
     cols_ = a.cols();
+    this->wait_for_read_write_events();
+    buffer_cl_ = a.buffer_cl_;
```

Comment:
This is also not OK. Copies of `cl::Buffer` are shallow - they point to same underlying memory on GPU. To get an actual deep copy that call to `enqueueCopyBuffer()` can not be avoided.

Reply:
`cl::Buffer` uses the `Wrapper` classes `operator=` (it uses `Memory::operator=` but that's just `Wrapper<cl_mem>::operator=`

```
    Wrapper<cl_type>& operator = (const Wrapper<cl_type>& rhs)
    {
        if (this != &rhs) {
            detail::errHandler(release(), __RELEASE_ERR);
            object_ = rhs.object_;
            detail::errHandler(retain(), __RETAIN_ERR);
        }
        return *this;
    }
```

is this a shallow copy? `object_` is just defined as

```
protected:
    cl_type object_;
```

which doesn't look like it would be shallow?

Reply:
Documentation for `cl::Buffer` https://github.khronos.org/OpenCL-CLHPP/classcl_1_1_buffer.html says:
> See Memory for details about copy semantics, etc.

Doc for Memory https://github.khronos.org/OpenCL-CLHPP/classcl_1_1_memory.html states:

> Copies of these objects are shallow, meaning that the copy will refer to the same underlying cl_mem as the original. For details, see clRetainMemObject() and clReleaseMemObject().

I guess that the object you are looking at is just a handle to underlying buffer.

Reply:
Ahh so now I get what your saying. But do we need a deep copy here? That assignment is going to call clReleaseMemObject() but that only releases the memory if the reference count is zero. But since the memory is being assigned the reference count should be greater than zero right?

Reply:
Deleting buffer here is not a problem, since we have anouther matrix that uses it. But we could have two matrix_cl objects that have same underlying memory. Well, we (you indluded) decided we want it to behave this way [here](https://github.com/stan-dev/math/issues/1307).

Reply:
Do the tests I added below for assignment cover what you are worried about? They check that a locally scoped matrix_cl's memory is correctly passed to an object defined outside of it's scope without deleting it. If those make sense then we can just use the one assign operator that uses forward

Feels silly not having these tests before!

https://github.com/stan-dev/math/pull/1344/files#diff-44c6cc026f22cf47e68bef5053262f4cR30

Reply:
As I said before deleting is NOT a problem. Having two matrices with the same buffer could be. If one is modifieds will be the other one.

Reply:
Totally forgot about that PR thanks for linking it, I'll fix these up to do the deep copies. I'll also add a test that shows when the shallow copy situation can break bad

Reply:
:+1 added a test for this so in the future if someone goes perfect forward loco they will be summarily stopped from goofing